### ISSUE TITLE
server/tes: call caching

### ISSUE THREAD

**Issue by buchanae** [ohsu-comp-bio/funnel#545]
I started experimenting with task call caching, and there are some big questions to answer. Wanted to start capturing some notes:

- It's best if the same task ID is returned for cached calls, ...or is it? I'm not sure, and this has a big effect on the items below.
- Should failed tasks be cached? What if it's a system error? Canceled?
- Should task input/output object ETags be included in the task cache key?
  This would significantly complicate things, since caching could not be completely
  handled in CreateTask, but would need an event writer and GetTask call.
- Will CreateTask need to make storage calls before returning? This could be a
  substantial hit to the speed of CreateTask.
- Docker image hashes are critical as well.

On one hand, it would be better if the costly calls to dockerhub (or other registry) and storage were async, handled by the compute.Submit call, but in that case the same task ID can't be returned for cached calls.

Perhaps this gets easier if a cached call is allowed to return a separate task ID, but I'm worried that clients relying on the cache might end up clogging up the task list with cached entries, since they'll be relying on the task to call CreateTask more frequently, less carefully. Perhaps a GetCachedTask endpoint could be added to allow a client to check whether a task is cached.

On the other hand, maybe a slower CreateTask call is acceptable and actually easier. Clients could adjust for this by calling CreateTask in parallel, and/or we could add a BulkCreateTask endpoint.
 
My (very) experimental code is here: https://github.com/buchanae/funnel/tree/call-caching

**Comment by buchanae** [comment]
Should creation/start/end times be updated in the logs for cached invocations? Is there a "cached" metadata field?

**Comment by kmavrommatis** [comment]
@buchanae 
has there been any recent activity on this?  Is there still interest to develop funnel to this direction?
This is a very interesting and important feature for a server than may be handling long running jobs.

**Comment by buchanae** [comment]
There hasn't been any activity on this, as far as I know. Yes, there's interest, just not enough time. It's a pretty complex feature.

**Comment by kellrott** [comment]
I feel like this is something that might better be served as a proxy server that sits in front of Funnel and identified potential repeat jobs before trying to run the job.
